src/blockchain/blockchain.service.ts

```diff
@@ -139,4 +139,85 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
+    return (await this.apiPromise.rpc.chain.getBlock()).block.header.number.toBigInt();
+  }
+
+  public async getLatestFinalizedBlockHash(): Promise<BlockHash> {
+    return (await this.apiPromise.rpc.chain.getFinalizedHead()) as BlockHash;
+  }
+
+  public async crawlBlockListForTx(
+    txHash: Hash,
+    blockList: bigint[],
+    successEvents: [{ pallet: string; event: string }],
+  ): Promise<{ found: boolean; success: boolean; blockHash?: BlockHash; capacityWithDrawn?: string; error?: RegistryError }> {
+    const txReceiptPromises: Promise<{ found: boolean; success: boolean; blockHash?: BlockHash; capacityWithDrawn?: string; error?: RegistryError }>[] = blockList.map(
+      async (blockNumber) => {
+        const blockHash = await this.getBlockHash(blockNumber);
+        const block = await this.getBlock(blockHash);
+        const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());
+
+        if (!txInfo) {
+          return { found: false, success: false };
+        }
+
+        this.logger.verbose(`Found tx ${txHash} in block ${blockNumber}`);
+        const at = await this.api.at(blockHash.toHex());
+        const eventsPromise = firstValueFrom(at.query.system.events());
+
+        let isTxSuccess = false;
+        let totalBlockCapacity: bigint = 0n;
+        let txError: RegistryError | undefined;
+
+        try {
+          const events = await eventsPromise;
+
+          events.forEach((record) => {
+            const { event } = record;
+            const eventName = event.section;
+            const { method } = event;
+            const { data } = event;
+            this.logger.debug(`Received event: ${eventName} ${method} ${data}`);
+
+            // find capacity withdrawn event
+            if (eventName.search('capacity') !== -1 && method.search('Withdrawn') !== -1) {
+              // allow lowercase constructor for eslint
+              // eslint-disable-next-line new-cap
+              const currentCapacity: u128 = new u128(this.api.registry, data[1]);
+              totalBlockCapacity += currentCapacity.toBigInt();
+            }
+
+            // check custom success events
+            if (successEvents.find((successEvent) => successEvent.pallet === eventName && successEvent.event === method)) {
+              this.logger.debug(`Found success event ${eventName} ${method}`);
+              isTxSuccess = true;
+            }
+
+            // check for system extrinsic failure
+            if (eventName.search('system') !== -1 && method.search('ExtrinsicFailed') !== -1) {
+              const dispatchError = data[0] as DispatchError;
+              const moduleThatErrored = dispatchError.asModule;
+              const moduleError = dispatchError.registry.findMetaError(moduleThatErrored);
+              txError = moduleError;
+              this.logger.error(`Extrinsic failed with error: ${JSON.stringify(moduleError)}`);
+            }
```

Comment:
This is getting all events of these types in the block; even if they were from txns submitted by other users, or for other txns. We need a way to filter to only the events related to the specifc tx hash.

Reply:
we filter based on tx hash

>    const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());

Reply:
No; we filter to see if the tx is in the block, but then we get all of the events in the block, regardless of which txn they belong to.

Reply:
so either we cache the blocks already pulled for any txs that fall within the range, or some sort of debouncer would do, but again not sure how much we need to do given the limited scope of this PR for now

Reply:
One more thing, even if we are doing 32 blocks in future, what is the probability that many blocks will have been created by the time the job is hit? we are looking ahead for blocks and not in past so majorly will hit only few blocks in ideal conditions

Reply:
In order to match an event to the extrinsic it belongs to, you need to compare the extrinsic index in the event to the index of the extrinsic from the transaction in the block. See here for one example of how to do this:
https://polkadot.js.org/docs/api/cookbook/blocks/#how-do-i-determine-if-an-extrinsic-succeededfailed